src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -66,169 +67,226 @@ public static string GetNuGetLockFilePath(string baseDirectory, string projectNa
-                {
-                    // lock file is out of sync
-                    return false;
-                }
-            }
+                    if (target == null)
+                    {
+                        // a new target found in the dgSpec so invalidate existing lock file.
+                        invalidReasons.Add(string.Format(
+                                    CultureInfo.CurrentCulture,
+                                    Strings.PackagesLockFile_NewTargetFramework,
+                                    framework.FrameworkName.GetShortFolderName())
+                                );
+
+                        continue;
+                    }
 
-            // Validate the runtimes for the current project did not change.
-            var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
-            var lockFileRuntimes = nuGetLockFile.Targets.Select(t => t.RuntimeIdentifier).Where(r => r != null).Distinct();
+                    IEnumerable<LockFileDependency> directDependencies = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.Direct);
 
-            if (!projectRuntimesKeys.OrderedEquals(
-                            lockFileRuntimes,
-                            x => x,
-                            StringComparer.InvariantCultureIgnoreCase,
-                            StringComparer.InvariantCultureIgnoreCase))
-            {
-                return false;
-            }
+                    (var hasProjectDependencyChanged, var pmessage) = HasProjectDependencyChanged(framework.Dependencies, directDependencies, target.TargetFramework);
+                    if (hasProjectDependencyChanged)
+                    {
+                        // lock file is out of sync
+                        invalidReasons.Add(pmessage);
+                    }
 
-            // Validate all P2P references
-            foreach (var framework in project.RestoreMetadata.TargetFrameworks)
-            {
-                var target = nuGetLockFile.Targets.FirstOrDefault(
-                    t => EqualityUtility.EqualsWithNullCheck(t.TargetFramework, framework.FrameworkName));
+                    var transitiveDependenciesEnforcedByCentralVersions = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.CentralTransitive).ToList();
+                    var transitiveDependencies = target.Dependencies.Where(dep => dep.Type == PackageDependencyType.Transitive).ToList();
 
-                if (target == null)
-                {
-                    // a new target found in the dgSpec so invalidate existing lock file.
-                    return false;
+                    (var hasTransitiveDependencyChanged, var tmessage) = HasProjectTransitiveDependencyChanged(framework.CentralPackageVersions, transitiveDependenciesEnforcedByCentralVersions, transitiveDependencies);
```

Comment:
There's this problem https://github.com/NuGet/Home/issues/10074...I don'timagine it's getting fixed here, but something to be aware of.